### PR TITLE
FIX: Destroy associated user api keys when making a user anonymous.

### PR DESCRIPTION
**Pull request by romanrizzi**

### CODE REVIEW

**Review by eviltrout — APPROVED** [review]

File: app/services/user_anonymizer.rb [hunk 1]
```diff
@@ -59,11 +59,12 @@ def make_anonymous
...
-      @user.user_avatar.try(:destroy)
-      @user.single_sign_on_record.try(:destroy)
-      @user.oauth2_user_infos.try(:destroy_all)
-      @user.user_associated_accounts.try(:destroy_all)
-      @user.api_keys.find_each { |x| x.try(:destroy) }
+      @user.user_avatar&.destroy
+      @user.single_sign_on_record&.destroy
+      @user.oauth2_user_infos&.destroy_all
+      @user.user_associated_accounts&.destroy_all
+      @user.api_keys.find_each { |x| x&.destroy }
```
[the remark this answers is not in the document]
Comment (romanrizzi) [reply app/services/user_anonymizer.rb:66]: I agree `destroy!` is better.  What do you think about wrapping all the saves and deletes inside a transaction?

[the remark this answers is not in the document]
Comment (tgxworld) [reply app/services/user_anonymizer.rb:66]: I think we already do :) https://github.com/discourse/discourse/blob/c2f13344292008bdfe4f533eb236a96d8bf10829/app/services/user_anonymizer.rb#L21

The entire block of code is wrapped in `User.transaction`

[the remark this answers is not in the document]
Comment (tgxworld) [reply app/services/user_anonymizer.rb:66]: Another thing to note is that if there are no `destroy` callbacks on the model, we can just use `destroy_all` which will delete the record directly from the database.

[the remark this answers is not in the document]
Comment (romanrizzi) [reply app/services/user_anonymizer.rb:66]: > I think we already do :) 

🤦 

> Another thing to note is that if there are no destroy callbacks on the model, we can just use destroy_all which will delete the record directly from the database.

👍

**Review by tgxworld — APPROVED** [review]

**Pull request merged**